testing/testing.h

```diff
@@ -28,18 +28,43 @@ struct test_t {
 
     const char* message() const
     {
-        return message_.c_str();
+        return message_.str().c_str();
     }
 
     bool has_message() const
     {
-        return !message_.empty();
+        return !message_.str().empty();
     }
 
 protected:
     bool pass_;
     const char* name_;
-    std::string message_;
+    std::stringstream message_;
 };
 
 void register_test(test_t* instance);
+
+#define TEST(GROUP, NAME)                           \
+    namespace {                                     \
+        struct GROUP##_##NAME##_t : public test_t { \
+            GROUP##_##NAME##_t()                    \
+                : test_t(#GROUP "." #NAME)          \
+            {                                       \
+                register_test(this);                \
+            }                                       \
+            virtual void run();                     \
+        } GROUP##_##NAME;                           \
+    }                                               \
+    void GROUP##_##NAME##_t::run()
+
+#define ASSERT(CONDITION)    \
+    if (!(CONDITION)) {      \
+        this->pass_ = false; \
+    } else {                 \
+        return;              \
+    }
+
+#define EXPECT(CONDITION)    \
+    if (!(CONDITION)) {      \
```

Comment:
Could wrap all of this inside a {} block to stop 'else' being able to associate with this macro.